java-client/session/src/main/java/io/deephaven/client/impl/BatchTableRequestBuilder.java

```diff
@@ -0,0 +1,582 @@
+            if (existing.isPresent()) {
+                return TableReference.newBuilder().setTicket(existing.get()).build();
+            }
+            final Integer ix = indices.get(table);
+            if (ix != null) {
+                return TableReference.newBuilder().setBatchOffset(ix).build();
+            }
+            throw new IllegalStateException(
+                "Unable to reference table - batch table request logic has a bug.");
+        }
+
+        @Override
+        public void visit(EmptyTable emptyTable) {
+            out = op(Builder::setEmptyTable,
+                EmptyTableRequest.newBuilder().setResultId(ticket).setSize(emptyTable.size()));
+        }
+
+        @Override
+        public void visit(NewTable newTable) {
+            throw new UnsupportedOperationException("TODO");
+        }
+
+        @Override
+        public void visit(TimeTable timeTable) {
+            // noinspection Convert2Lambda
+            timeTable.timeProvider().walk(new Visitor() {
+                @Override
+                public void visit(TimeProviderSystem system) {
+                    // OK
+                    // note: if other time providers are added, we need to throw an exception
+                }
+            });
+
+            TimeTableRequest.Builder builder = TimeTableRequest.newBuilder().setResultId(ticket)
+                .setPeriodNanos(timeTable.interval().toNanos());
+            if (timeTable.startTime().isPresent()) {
+                final Instant startTime = timeTable.startTime().get();
+                final long epochNanos = Math.addExact(
+                    TimeUnit.SECONDS.toNanos(startTime.getEpochSecond()), startTime.getNano());
+                builder.setStartTimeNanos(epochNanos);
+            }
+            out = op(Builder::setTimeTable, builder.build());
+        }
+
+        @Override
+        public void visit(MergeTable mergeTable) {
+            MergeTablesRequest.Builder builder =
+                MergeTablesRequest.newBuilder().setResultId(ticket);
+            for (TableSpec table : mergeTable.tables()) {
+                builder.addSourceIds(ref(table));
+            }
+            out = op(Builder::setMerge, builder.build());
+        }
+
+        @Override
+        public void visit(HeadTable headTable) {
+            out = op(Builder::setHead, HeadOrTailRequest.newBuilder().setResultId(ticket)
+                .setSourceId(ref(headTable.parent())).setNumRows(headTable.size()));
+        }
+
+        @Override
+        public void visit(TailTable tailTable) {
+            out = op(Builder::setTail, HeadOrTailRequest.newBuilder().setResultId(ticket)
+                .setSourceId(ref(tailTable.parent())).setNumRows(tailTable.size()));
+        }
+
+        @Override
+        public void visit(ReverseTable reverseTable) {
+            // a bit hacky at the proto level, but this is how to specify a reverse
+            out = op(Builder::setSort,
```

Comment:
Are we planning on visiting any of the easy tree optimizations? Such as, here if reverseTable.parent() is also a ReverseTable then you can simply return the `ref(reverseTable.parent().parent())` (with appropriate casting). I believe the web has (or had) a couple of quick and dirty optimizations.

Or, should we consider taking the batch request server side and converting to a QST that we can run over such an optimizer? Maybe this is best .. and it probably is possible to rewrite the batch processing in the table service to use QST visitors. Do you know if we already have a ticket for this?

Opinions?

Reply:
I think dag optimizations belong at the QST level, not at the individual visitor level. This way, all visitors benefit, instead of just some. It would be very easy to create a pluggable framework where you could choose which optimizations you want to apply (the most common use case being all of them; but I can see some use cases where you don't want all, or you want to add custom ones).